### PR TITLE
Add speedometer options

### PR DESCRIPTION
**Pull request by Gaberboo**
Adds options for different speedometer metrics

### CODE REVIEW

File: assets/message/MenuSP_U.bmg.json5 [hunk 1]
```diff
@@ -1095,6 +1095,50 @@
...
         "string": "No track was found in the\n/mkw-sp/thumbnails folder.",
     },
+    "10315": {
+        "font": "regular",
+        "string": "Speedometer",
+    },
+    "10316": {
+        "font": "regular",
+        "string": "Internal + Terrain",
+    },
```
Comment (vabold) [inline assets/message/MenuSP_U.bmg.json5:1105]: If possible, a better name could be used. "Terrain" implies more general KCL, not just moving water/road.

File: payload/game/ui/ctrl/CtrlRaceSpeed.cc [hunk 2]
```diff
@@ -53,24 +54,52 @@ void CtrlRaceSpeed::calcSelf() {
...
+                    vel = Vec3{ pos->x - lastPos->x, 0.0f , pos->z - lastPos->z };
+                } else if (setting == SP::ClientSettings::Speedometer::Y) {
+                    vel = Vec3{ 0.0f, pos->y - lastPos->y, 0.0f };
+                }
+                speed = PSVECMag(&vel);
+                break;
+            } 
+    }
+    f32 hardSpeedLimit = object->getKartMove()->hardSpeedLimit();
+    speed = std::min(speed, hardSpeedLimit);
```
Comment (vabold) [inline payload/game/ui/ctrl/CtrlRaceSpeed.cc:101]: Hard speed limit should not be applied to XYZ/XZ/Y, given that we are measuring more than the velocity.

File: payload/game/ui/ctrl/CtrlRaceSpeed.cc [hunk 3]
```diff
@@ -53,24 +54,52 @@ void CtrlRaceSpeed::calcSelf() {
...
-        Vec3 vel{ pos->x - lastPos->x, pos->y - lastPos->y, pos->z - lastPos->z };
-        speed = PSVECMag(&vel);
+        setting = SP::ClientSettings::Speedometer::XYZ;
     }
-
+    f32 speed = 0.0f;
+    switch (setting) {
+        case SP::ClientSettings::Speedometer::InternalPlus:
+        case SP::ClientSettings::Speedometer::Internal:
+            {
```
Comment (vabold) [inline payload/game/ui/ctrl/CtrlRaceSpeed.cc:70]: We don't use braces for code in switch cases.

File: payload/game/ui/ctrl/CtrlRaceSpeed.cc [hunk 4]
```diff
@@ -53,24 +54,52 @@ void CtrlRaceSpeed::calcSelf() {
...
+                    const Vec3 *movingWaterVel = object->getVehiclePhysics()->movingWaterVel();
+                    speed += PSVECDotProduct(internalVelDir, movingRoadVel);
+                    speed += PSVECDotProduct(internalVelDir, movingWaterVel);
+                }
+                break;
+            }
+        case SP::ClientSettings::Speedometer::XYZ:
+        case SP::ClientSettings::Speedometer::XZ:
+        case SP::ClientSettings::Speedometer::Y:
+            {
```
Comment (vabold) [inline payload/game/ui/ctrl/CtrlRaceSpeed.cc:85]: Same as above.

**Review by vabold — CHANGES_REQUESTED** [review]

File: assets/message/MenuSP_U.bmg.json5 [hunk 5]
```diff
@@ -1118,6 +1118,50 @@
     "10320": {
         "font": "regular",
         "string": "Standard checkpoint overlay.",
+    }
```
Comment (vabold) [inline assets/message/MenuSP_U.bmg.json5:1121]: `},`

File: payload/game/ui/ctrl/CtrlRaceSpeed.cc (hunk 3 above)
Comment (riidefi) [reply payload/game/ui/ctrl/CtrlRaceSpeed.cc:70]: It's needed to declare variables inside the scope. Should be fine as long as he runs clang-format here and fixes the indentation of the case labels/such.

File: payload/game/ui/ctrl/CtrlRaceSpeed.cc [hunk 6]
```diff
@@ -53,22 +54,48 @@ void CtrlRaceSpeed::calcSelf() {
...
-    const Vec3 *movingRoadVel = object->getVehiclePhysics()->movingRoadVel();
-    const Vec3 *movingWaterVel = object->getVehiclePhysics()->movingWaterVel();
-    f32 speed = internalSpeed;
-    speed += PSVECDotProduct(internalVelDir, movingRoadVel);
-    speed += PSVECDotProduct(internalVelDir, movingWaterVel);
-    f32 hardSpeedLimit = object->getKartMove()->hardSpeedLimit();
-    speed = std::min(speed, hardSpeedLimit);
 
     if (object->getKartState()->inCannon()) {
+        setting = SP::ClientSettings::Speedometer::XYZ;
```
Comment (stblr) [inline payload/game/ui/ctrl/CtrlRaceSpeed.cc:64]: Do we want this for all settings?

File: assets/message/MenuSP_U.bmg.json5 [hunk 7]
```diff
@@ -1119,6 +1119,50 @@
         "font": "regular",
         "string": "Standard checkpoint overlay.",
     },
+    "10321": {
+        "font": "regular",
+        "string": "Speedometer",
+    },
+    "10322": {
+        "font": "regular",
+        "string": "Internal + Moving Road",
```
[the remark this answers is not in the document]
Comment (Gaberboo) [reply assets/message/MenuSP_U.bmg.json5:1128]: I just liked moving road better

File: payload/game/ui/ctrl/CtrlRaceSpeed.cc (hunk 6 above)
Comment (Gaberboo) [reply payload/game/ui/ctrl/CtrlRaceSpeed.cc:64]: Yes

**Review by stblr — APPROVED** [review]

**Pull request merged**